apollo-federation/src/operation/integrity.rs

```diff
@@ -0,0 +1,202 @@
+                        fragment_data.type_condition_position, fragment_def.type_condition_position
+                    )));
+                }
+
+                if matches!(option, IsWellFormedOption::CheckFragmentSpreadSelectionSet) {
+                    fragment_spread.selection_set.is_well_formed(
+                        schema,
+                        named_fragments,
+                        option,
+                    )?;
+                }
+                Ok(())
+                // Note: fragment_data.type_condition_position and the parent type do not have to have
+                // non-empty intersection to be well-formed. It would be an extra check.
+            }
+            Selection::InlineFragment(inline_fragment) => {
+                let fragment_data = inline_fragment.inline_fragment.data();
+                if fragment_data.schema != *schema {
+                    return Err(FederationError::internal(format!(
+                        "Schema mismatch: expected {:?}, got {:?}",
+                        schema, fragment_data.schema
+                    )));
+                }
+                if fragment_data.parent_type_position != *parent_type {
+                    return Err(FederationError::internal(format!(
+                        "Parent type mismatch: expected {:?}, got {:?}",
+                        parent_type, fragment_data.parent_type_position
+                    )));
+                }
+                if fragment_data.casted_type() != inline_fragment.selection_set.type_position {
+                    return Err(FederationError::internal(format!(
+                        "Inline fragment's casted-type ({:?}) and the type of its sub-selection-set ({:?}) mismatch.",
+                        fragment_data.casted_type(),
+                        inline_fragment.selection_set.type_position
+                    )));
+                }
+                inline_fragment
+                    .selection_set
+                    .is_well_formed(schema, named_fragments, option)
+                // Note: fragment_data.type_condition_position and the parent type do not have to have
+                // non-empty intersection to be well-formed. It would be an extra check.
+            }
+        }
+    }
+}
+
+impl SelectionSet {
+    pub fn is_well_formed(
+        &self,
+        schema: &ValidFederationSchema,
+        named_fragments: &NamedFragments,
+        option: IsWellFormedOption,
+    ) -> Result<(), FederationError> {
+        if self.schema != *schema {
+            return Err(FederationError::internal(format!(
+                "Schema mismatch: expected {:?}, got {:?}",
+                schema, self.schema
+            )));
+        }
+        for selection in self.iter() {
+            selection.is_well_formed(schema, named_fragments, &self.type_position, option)?;
+        }
+        Ok(())
+    }
+}
+
+impl Operation {
+    pub fn is_well_formed(
```

Comment:
I'm noticing that:
1. The well-formedness check for fragment spread selections recurses into the fragment's selection sets, each time the fragment spread is referenced (assuming they're not skipped).
2. We don't check the well-formedness of `Operation.named_fragments` here. I'm guessing because we assume they should be exactly the same as the fragments referenced above (although we're not checking this here).

Looking at the codebase, the `FragmentSpreadSelection.selection_set` field, when directly set, comes from existing `Fragment`s. That means that it should always match the corresponding `Fragment`'s selection set in the operation. And not just `Eq` here for `SelectionSet`, but more specifically `SelectionSet.schema` and `SelectionSet.type_position` should be `Eq`, and `SelectionSet.selections` should compare the same using `Arc::ptr_eq()`.

This would suggest that we change the above validation to:
1. Validate the fragments in `Operation.named_fragments` (recursing into their selection sets, but also e.g. checking that `Fragment.type_condition_position` matches the one in the selection set).
2. When validating a fragment spread selection, we should check that the `FragmentSpreadSelection.selection_set` matches in the way described above.

There's a similar case to be made for `FragmentSpreadSelection.spread`, where it always comes from feeding a `Fragment` to `FragmentSpreadData::from_fragment()`. This means we should similarly check matching there, i.e. we should check for `Eq` for `FragmentSpreadData.schema`, `FragmentSpreadData.fragment_name`, and `FragmentSpreadData.type_condition_position`, and check `Arc::ptr_eq()` for `FragmentSpreadData::fragment_directives()`.
- While looking here, I noticed a bug in `FragmentSpreadSelection::new()` where it re-fetches its fragment selection set from `NamedFragments`, but doesn't update the spread data. Glancing at the code, this looks to be used only for `SelectionSet::make_selection()`, specifically in `Selection::from_operation_element()`. The latter function appears to be a rename of `selectionOfElement()` from JS, but the JS code converts fragment spreads to inline fragments, while the Rust code appears to be keeping them as fragment spreads. It would be better I think here to copy the JS code's behavior (or error when fragment spreads are present, if we believe they shouldn't occur).